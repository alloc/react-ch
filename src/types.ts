export type ChannelEffect<T = any, U = any> = (
  ...args: Tuple<T>
) => U | Promise<U>

export type DisposableEffect<T = any, U = any> = {
  effect: ChannelEffect<T, U>
  dispose: () => void
}

export type AsyncFunction = (...args: any[]) => Promise<any>

/**
 * Get a tuple type from any type.
 *
 * - The `any` type becomes the `any[]` type.
 * - The `void` and `never` types become the `[]` type.
 * - Finite array types (aka tuples) stay the same.
 * - All other types become the `[T]` type.
 */
export type Tuple<T = any> = unknown &
  ([T] extends [never]
    ? []
    : [T] extends [Any]
    ? any[]
    : [T] extends [void]
    ? []
    : T extends ReadonlyArray<any>
    ? T[number][] extends T
      ? [T]
      : T
    : [T])

// Used for "any" detection.
declare class Any {
  private _: Any
}
