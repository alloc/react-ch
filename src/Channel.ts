import is from '@alloc/is'
import { AsyncFunction, ChannelEffect, DisposableEffect, Tuple } from './types'

export interface Channel<T = any, U = any> {
  (...args: Tuple<T>): Promise<U[]>
  once(effect: ChannelEffect<T, U>): DisposableEffect<T, U>
}

/**
 * Channels allow for loose coupling between a function call and its effect(s).
 *
 * Its `T` type is either the first and only argument type passed to the channel
 * (eg: `T = number`), or an array type representing varargs of the channel
 * (eg: `T = [number, string] | number[]`).
 *
 * Its `U` type is the expected return type of any effect functions that react
 * to the channel (eg: `U = Promise<void> | void`).
 */
export class Channel<T = any, U = any> {
  /** The channel name. Useful for debugging. */
  readonly name!: string
  protected effects!: Set<ChannelEffect<T, U>>

  /** Override this to wrap every channel emitter. */
  static wrapEmit: AsyncFunction = emit => emit

  constructor(name?: string, effect?: ChannelEffect<T, U>)
  constructor(effect: ChannelEffect<T, U>)
  constructor(arg1?: string | ChannelEffect, arg2?: ChannelEffect) {
    const effects = new Set<ChannelEffect>()
    const channel: any = Channel.wrapEmit((...args: Tuple<T>) =>
      Promise.all(Array.from(effects, effect => effect(...args)))
    )

    Object.setPrototypeOf(channel, Channel.prototype)
    if (arg1) {
      if (is.string(arg1)) {
        Object.defineProperty(channel, 'name', { value: name })
      } else {
        effects.add(arg1)
      }
    }
    if (arg2) {
      effects.add(arg2)
    }
    channel.effects = effects
    return channel
  }

  /** Listen until disposed. */
  on(effect: ChannelEffect<T, U>): DisposableEffect<T, U> {
    this.effects.add(effect)
    return {
      effect,
      dispose: () => {
        this.effects.delete(effect)
      },
    }
  }

  /** Listen for the next call or until disposed. */
  once(effect: ChannelEffect<T, U>): DisposableEffect<T, U> {
    const once = this.on(
      (...args: any): any => (once.dispose(), effect(...args))
    )
    once.effect = effect
    return once
  }
}

const ch = new Channel()
ch.name
