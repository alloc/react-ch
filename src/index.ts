// Hooks
export { useChannel } from './useChannel'

// Classes
export { Channel } from './Channel'

// Types
export { ChannelEffect, DisposableEffect } from './types'
