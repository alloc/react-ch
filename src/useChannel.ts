import { is } from '@alloc/is'
import React from 'react'
import { useLayoutEffect } from 'react-layout-effect'
import { Channel } from './Channel'
import { ChannelEffect } from './types'

const noop = Function.prototype

/** Add an effect to a channel */
export function useChannel<T, U = void>(
  channel: Channel<T, U> | undefined,
  effect: ChannelEffect<T, U> | undefined,
  deps?: any[]
): void

/** Create a channel */
export function useChannel<T, U = void>(name?: string): Channel<T, U>

/** Create a channel with an effect */
export function useChannel<T, U = void>(
  effect: ChannelEffect<T, U>,
  deps?: any[]
): Channel<T, U>

/** Create a channel with an effect */
export function useChannel<T, U = void>(
  name: string,
  effect: ChannelEffect<T, U> | undefined,
  deps?: any[]
): Channel<T, U>

/** @internal */
export function useChannel(
  arg1?: string | Channel | ChannelEffect,
  arg2?: ChannelEffect | any[],
  arg3?: any[]
) {
  const name = is.string(arg1) ? arg1 : ''

  const channel =
    arguments.length > 1 && (arg1 == null || arg1 instanceof Channel)
      ? arg1
      : React.useState(() => new Channel(name))[0]

  const effect: any =
    channel && arg1 !== channel && is.function(arg1) ? arg1 : arg2

  if (arguments.length > 1 || (effect && arg1 == effect)) {
    // Replace the effect without changing call order.
    const effectRef = React.useRef<ChannelEffect>(effect)
    useLayoutEffect(
      () => {
        effectRef.current = effect || noop
      },
      // The deps array decides when the effect is replaced.
      is.array(arg2) ? arg2 : arg3
    )

    // Start listening on commit, and stop on unmount.
    useLayoutEffect(
      () =>
        channel && channel.on((...args) => effectRef.current(...args)).dispose,
      [channel]
    )
  }

  // Return the channel only if we created it.
  if (arg1 !== channel) {
    return channel
  }
}
