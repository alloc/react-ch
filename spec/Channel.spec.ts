import { is } from '@alloc/is'
import { flushMicroTasks } from 'flush-microtasks'
import { describe, expect, it, Mock, vi } from 'vitest'
import { Channel } from '../src/Channel'

const getCalls = (channel: Channel<any>) =>
  Array.from(
    channel['effects'] as Set<Mock>,
    effect => effect.mock && effect.mock.calls
  ).filter(Boolean)

describe('Channel', () => {
  it('can pass variable arguments to effect functions', () => {
    const test = new Channel<[number, number]>()
    test.on(vi.fn())
    test.on(vi.fn())
    test(1, 2)
    test(3, 4)
    expect(getCalls(test)).toMatchSnapshot()
  })

  it('can pass nothing to effect functions', () => {
    const test = new Channel<void>()
    test.on(vi.fn())
    test.on(vi.fn())
    test()
    test()
    expect(getCalls(test)).toMatchSnapshot()
  })

  it('can remove an effect during emit', () => {
    const test = new Channel()
    const sub = test.on(() => sub.dispose())
    test.on(vi.fn())
    test()
    expect(getCalls(test)).toMatchSnapshot()
  })

  describe('when no effects exist', () => {
    it('resolves with an empty array', async () => {
      const test = new Channel()
      const promise = test()
      expect(is.promise(promise)).toBeTruthy()
      expect(await promise).toEqual([])
    })
  })

  describe('when an effect is added with .once()', () => {
    it('only calls the effect once', () => {
      const test = new Channel()
      test.once(vi.fn())
      expect(getCalls(test)).toMatchSnapshot()
    })
  })

  describe('when effects return promises', () => {
    it('wraps those promises with Promise.all', async () => {
      const test = new Channel<void, number>()
      const promises: { resolve(): void }[] = []
      const createPromise = <T>(value: T) =>
        new Promise<T>(resolve =>
          promises.push({
            resolve: () => resolve(value),
          })
        )

      test.on(() => createPromise(1))
      test.on(() => createPromise(2))

      let result: any[] | undefined
      test().then(value => {
        result = value
      })

      await flushMicroTasks()
      expect(result).toBeUndefined()

      promises.forEach(p => p.resolve())

      await flushMicroTasks()
      expect(result).toEqual([1, 2])
    })
  })
})
