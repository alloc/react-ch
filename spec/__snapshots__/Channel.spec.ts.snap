// Vitest Snapshot v1

exports[`Channel > can pass nothing to effect functions 1`] = `
[
  [
    [],
    [],
  ],
  [
    [],
    [],
  ],
]
`;

exports[`Channel > can pass variable arguments to effect functions 1`] = `
[
  [
    [
      1,
      2,
    ],
    [
      3,
      4,
    ],
  ],
  [
    [
      1,
      2,
    ],
    [
      3,
      4,
    ],
  ],
]
`;

exports[`Channel > can remove an effect during emit 1`] = `
[
  [
    [],
  ],
]
`;

exports[`Channel > when an effect is added with .once() > only calls the effect once 1`] = `[]`;
